glean-core/ios/Glean/Scheduler/MetricsPingScheduler.swift

```diff
@@ -0,0 +1,216 @@
+    struct Constants {
+        static let logTag = "glean/MetricsPingSched"
+        static let lastMetricsPingSentDateTime = "last_metrics_ping_iso_datetime"
+        static let dueHourOfTheDay = 4
+        static let lastVersionOfAppUsed = "last_version_of_app_used"
+    }
+
+    private let logger = Logger(tag: Constants.logTag)
+
+    var timer: Timer?
+
+    /// Schedules the metrics ping collection at the due time.
+    ///
+    /// - parameters:
+    ///     * now:  A `Date` representing the current date/time
+    ///     * sendTheNextCalendarDay:  Determines whether to schedule collection for the next calendar day
+    ///                            or to attempt to schedule it for the current calendar day. If the latter and
+    ///                            we're overdue for the expected collection time, the task is scheduled for
+    ///                            immediate execution.
+    func schedulePingCollection(_ now: Date, sendTheNextCalendarDay: Bool) {
+        var fireDate = Calendar.current.date(
+            bySettingHour: Constants.dueHourOfTheDay,
+            minute: 0,
+            second: 0,
+            of: now
+        )!
+
+        // Invalidiate the timer if it's already running
+        timer?.invalidate()
+
+        // Setup timer and schedule
+        if sendTheNextCalendarDay {
+            let tomorrow = Calendar.current.date(
+                byAdding: .day,
+                value: 1,
+                to: fireDate,
+                wrappingComponents: true
+            )!
+            fireDate = tomorrow
+        }
+
+        logger.debug(
+            "Scheduling the 'metrics' ping for \(fireDate), in \(fireDate - now) seconds."
+        )
+
+        // Set the timer to fire at the `fireDate`
+        timer = Timer.scheduledTimer(withTimeInterval: fireDate - now, repeats: false) { _ in
+            self.logger.debug("MetricsPingScheduler timer fired!")
+            // When the timer fires, call `collectPingAndReschedule` with the current
+            // date/time.
+            self.collectPingAndReschedule(Date(), startupPing: false)
+        }
+    }
+
+    /// Determines if the application is a differnet version from the last time it was run. This is used to prevent
+    /// mixing data from multiple versions of the application in the same ping.
+    ///
+    /// - returns: `true` if the version is different, `false` if the version is the same.
+    func isDifferentVersion() -> Bool {
+        // Determine if the version has changed since the last time we ran.
+        let currentVersion = AppInfo.displayVersion
+        if let lastVersion = UserDefaults.standard.string(forKey: Constants.lastVersionOfAppUsed) {
+            if currentVersion != lastVersion {
+                UserDefaults.standard.set(currentVersion, forKey: Constants.lastVersionOfAppUsed)
+                return true
+            }
+        }
```

Comment:
We should probably sent an update ping if we can't determine the earlier version, to match the logic in Android.  This only sends one if it *can* load the old version *and* it doesn't match.